### DISCUSSION TITLE
localStorage is not defined - Next.js

### DISCUSSION THREAD

**Discussion by danilo-leal** [vercel/next.js#19911] [Help] [answered]
Hey y'all.

This is about my first time trying out Next.js. I'm not really much of a coder, just trying to study some stuff out. I've set up a dark mode function, following [this video right here](https://www.youtube.com/watch?v=2IfTD-muRF4&ab_channel=DailyWebCoding). At first, it was working normally. But all of the sudden, it started popping up this localStorage error and I don't know what to do.

I'm working with Tailwind CSS too. This is the function:

```
import React, { useEffect, useState } from 'react';

export default function DarkModeToggle() {
  const [theme, setTheme] = useState(localStorage.theme);
  const colorTheme = theme === 'dark' ? 'light' : 'dark';
  useEffect(() => {
    const root = window.document.documentElement;

    root.classList.remove(colorTheme);
    root.classList.add(theme);
    localStorage.setItem('theme', theme);
  }, [theme, colorTheme]);

  return [colorTheme, setTheme];
}
```

Would love some help here.
Thanks a lot.

**Comment by jamesmosier** [comment, marked as answer]
Hi @danilo-leal. The issue you are seeing is because `localStorage` (aka `window.localStorage`) is not defined on the server side. Next server renders your components, so when that happens and it tried to access `localStorage` it can't find it. You'll have to wait until the browser renders it in order to use `localStorage`.

For your example, you should just be able to change `const [theme, setTheme] = useState(localStorage.theme);` to `const [theme, setTheme] = useState(null);` because your `useEffect` will run once React renders your component on the client side (`useEffect` only uses on the client, not server) and then `localStorage` will have the `theme` value after that effect runs.

---

For other use cases you might have that use `window in the future, you can check `if (typeof window !== 'undefined')` which is a good way to check for whether you are on the client vs server.

Or another approach would be to use Next's [dynamic imports with `ssr: false`](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr) although this is usually only needed for more involved use cases (i.e. if you are importing a chart/graph library or something similar that uses `window`).

**Reply by danilo-leal** [reply]
Worked like a charm. Thanks for illuminating here!

**Reply by danilo-leal** [reply]
@jamesmosier 
Hey! Revisiting this one because I have a sense that defining that `useState` to null made the browser not remember what theme was on when changing routes. Any idea whats going on? I'm clueless here. You can take a peak at the repo in my profile!

**Reply by adebiyial** [reply]
Instead of setting the state to `null`, have you tried defaulting it to "dark" or "light"?

**Reply by danilo-leal** [reply]
I get a **_ReferenceError: light is not defined_** when passing the light mode in `useState(light)`

**Reply by jamesmosier** [reply]
You'd want that to be a string: `useState('light')`.

Aside from that, here are a couple good blogs posts on this problem:

https://brianlovin.com/overthought/adding-dark-mode-with-next-js
https://www.joshwcomeau.com/react/dark-mode/

And there is also this discussion which touches on it: https://github.com/vercel/next.js/discussions/12171

---

After looking at your repository, I'd recommend putting your navbar (with the dark mode toggle) in your [_app.js](https://github.com/danilo-leal/daniloleal.co/blob/main/pages/_app.js). This will then persist (i.e. it will not be re-rendered) the changes as you navigate from one page to another.

**Reply by danilo-leal** [reply]
@jamesmosier 

Thanks for the attention, dude. I have two components holding the DarkModeButton, which toggles the function. I have the NavBar and an AppBar. How should go I about these? Should put only the DarkModeButton in the _app.js so then the theme persists across all pages and components holding it? The NavBar stays mostly at the home page while the AppBar is in all sub-pages.

By setting the theme in the useState, the browser don't hold the current theme and always goes back to the one set in the useState. I deployed it here if you want to take a look: https://dl-co.vercel.app/

**Reply by adebiyial** [reply]
@danilo-leal Ideally, you'd be using a global state through Context for example. I have this kind of stuff on my Next.js website (https://orangeman.dev), but here are also two articles to help guide you:

1. https://epicreact.dev/css-variables/
2. https://css-tricks.com/a-dark-mode-toggle-with-react-and-themeprovider/

If you still can't find a way to fix this, we can have a screen-sharing session to fix it once and for all.

**Reply by danilo-leal** [reply]
@adebiyial I think my main problem rests in the dark mode toggle function. I'm using TailwindCSS built-in feature for toggling between themes. It consists in adding a `dark:` prefix to the utility classes. I'm afraid there's still something wrong with the localStorage stuff, since the browser isn't remembering the current selected theme when changing routes.

But hey, thanks so much for all the attention here!

**Reply by adebiyial** [reply]
Hmm. Interesting. I used styled-components. Nevertheless, try this

```
import React, { useEffect, useState } from 'react';

export default function DarkModeToggle() {
  const [theme, setTheme] = useState(localStorage.theme);

  useEffect(() => {
    // Move this in
    const colorTheme = theme === 'dark' ? 'light' : 'dark';
    const root = window.document.documentElement;

    root.classList.remove(colorTheme);
    root.classList.add(theme);
    localStorage.setItem('theme', theme);
  }, [theme, colorTheme]);

  // return `theme` instead of `colorTheme`
  return [theme, setTheme];
}
```

**Reply by danilo-leal** [reply]
`ReferenceError: localStorage is not defined` :( 

There's something weird happening that I can't use the localStorage in the useState there.

**Reply by adebiyial** [reply]
That's where you should set your default theme. Don't worry, Next.js will run the `useEffect()` and present the user with the updated theme

Try this.

```
export default function DarkModeToggle() {
  const [theme, setTheme] = useState("dark");
 
  useEffect(() => {
    // will be "dark" or "light"
    const colorTheme = localStorage.getItem("theme");

   const root = window.document.documentElement;
 
   // Walk your way up from here...
   root.classList.replace(colorTheme, theme);
   localStorage.setItem('theme', theme);
  }, [theme, colorTheme]);

  // return `theme` instead of `colorTheme`
  return [theme, setTheme];
}
```

**Reply by danilo-leal** [reply]
I'm now seeing an error pointing out that `colorTheme` hasn't been defined 🤔

**Reply by adebiyial** [reply]
OK. Do something like

`const colorTheme = localStorage.getItem("theme") || theme`

That will default to the theme in case there's no theme set in the local storage.

**Reply by danilo-leal** [reply]
Not working :( The toggle isn't functioning anymore.

**Reply by adebiyial** [reply]
Hmm. OK. I think you should recreate this on codesandbox or codespaces, so there can be a practical debugging.

**Reply by danilo-leal** [reply]
I'm not sure if I made it right nor if it's working properly, but I've sorta uploaded to a Sandbox: https://codesandbox.io/s/trusting-haslett-f31wk

**Reply by vital-tech-results** [reply]
thank you @jamesmosier this code ```For other use cases you might have that use window in the future, you can check if (typeof window !== 'undefined')` which is a good way to check for whether you are on the client vs server.``` 

worked for me. 

I am using [appy theme by Cruip](https://cruip.com/demos/appy/) which incorporates tailwindCSS new "dark-mode" option.

**Reply by danilo-leal** [reply]
hey guys!

just coming back to this thread for a bit to share a blog post of mine where I go through how I ended up implementing the dark mode, which made me come here to take some questions by the time I was battling out on how to do it. finally got to a very good design, id like for you to check it out :)

https://daniloleal.co/my-world/thinking/darkmode-next-tailwind

**Reply by adebiyial** [reply]
This is great. I'm glad you were able to figure it out. Kudos!

Ps: I read through and found that and this typo "espeacially"

**Reply by blyscuit** [reply]
I used
```
const [theme, setTheme] = useState(
    typeof window !== "undefined" ? localStorage.theme : "light"
  );
```
and some component will show the wrong theme color when refresh the page. Use this answer and everything works perfectly.

**Reply by developerEhsan** [reply]
hy @jamesmosier this is working well on dev mode using dynamic import & "use client" but it doesn't allow me to create production ready website

**Reply by razafarishta** [reply]
Hi @Your-Ehsan 
Did you get any solution related to this?

**Reply by developerEhsan** [reply]
@razafarishta yes, solution is simple for the next.js I am using next-theme package 📦  which is specially built for next.js
you can take guidance from here https://github.com/pacocoursey/next-themes

this is also compatible with tailwindcss
so let me know if you find this helpfull

**Reply by razafarishta** [reply]
@Your-Ehsan Thanks for your reply
The problem I am facing is that I have used sessionStorage in my project, locally it is working fine but when I am deploying my project on Vercel it throws an error that sessionStorage is not defined. If you know the solution for this it will be very useful for me.
I am working on the next js V 14.
Thanks.

**Reply by developerEhsan** [reply]
@razafarishta I Think storing theme details on user's session storage is not recommended by the industry line, As it is not considered as a good practice.
But if you really want to do this in this way then you should check zustand with persistent using session storage 

check it out
https://github.com/pmndrs/zustand?tab=readme-ov-file#persist-middleware

**Reply by razafarishta** [reply]
@Your-Ehsan Thanks for the help
I will try it to implement using Zustand. 
Thanks

**Comment by IlirBajrami** [comment]
This discussion helped me a lot on understanding how localsotrage works in Next js. However i'm having a more specific case. 
I'm converting an app which used express into Next. I used `if (typeof window !== "undefined")` in useEffect hook but i have this in my useState: `const [job, setJob] = useState( localStorage.getItem("job") ? JSON.parse(localStorage.getItem("job")) : null );` 
and i'm getting `ReferenceError: localStorage is not defined` here.

Anyone have any clue how do i fix this?

**Reply by richardnguyen99** [reply]
I think it's the same as the answer above. `window.localStorage` will be available after `nextjs` rendered your app. That's why you can't use `localStorage` in `useState` because at that stage, it's on the Next server's job, which doesn't have `localStorage`.

**Comment by kiennguyen278** [comment]
I used:
`const ISSERVER = typeof window === "undefined";`
`if (!ISSERVER)
    localStorage.setItem(key, value);`
and it worked for me
hope it help you

**Reply by DevBrasil** [reply]
Thanks, help me alot.

**Reply by Sooraj-Rao** [reply]
Thanks, build was failing regularly after this check it got resolved!

**Reply by himanshu-tyd** [reply]
This did work for me thank you bro

**Comment by raheelshan** [comment]
You can use this code snippet.
`if (process.browser) {
            if (localStorage != undefined) {
                const token = localStorage.getItem("token");

                if (token != null) {
                    headers = {
                        Accept: "application/json",
                        Authorization: `Bearer ${token}`,
                    }
                }
            }
        }`

**Reply by ggepenyan** [reply]
what kind of variable is that 'headers'? Is that inside the configs? Is there any way to put something like that inside the configs? Or any way to do that after getting the token? I mean to set a global header.

**Comment by alexotoo** [comment]
**this is how I will do it using lazy initialization:********

import React, { useEffect, useState } from 'react';

const getInitialTheme =()=>window.localStorage.getItem('theme')

export default function DarkModeToggle() {
  const [theme, setTheme] = useState(getInitialTheme);
  const colorTheme = theme === 'dark' ? 'light' : 'dark';
  useEffect(() => {
    const root = window.document.documentElement;

    root.classList.remove(colorTheme);
    root.classList.add(theme);
    localStorage.setItem('theme', theme);
  }, [theme, colorTheme]);

  return [colorTheme, setTheme];
}

**Comment by kt-anas** [comment]
does this work in server-side
